### ISSUE TITLE
I want to be able to run, and build Linux and Windows (LCOW) containers simultaneously. 

### ISSUE THREAD

**Issue by bplasmeijer** [docker/roadmap#79]
**Tell us about your request**
I want to be able to run and build Linux and Windows (LCOW) containers simultaneously. 

**Which service(s) is this request for?**
docker build, and docker-compose

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
At this moment, I am using the (LCOW) **experimental feature** to run Linux and Windows containers simultaneously. The option to work streamless with Linux and Windows container is hard, because of the switch context option. 

**Are you currently working around the issue?**
- Switch context when building Linux images.
- Switch setting back to windows containers, and pull Linux image in (LCOW) mode.

**Additional context**
The new 20H1 WSL docker integration is a great new feature, and I love it, but integration requires Linux mode. I do lose my windows context. 
Issue #78 describes building both Linux and windows images in parallel on windows.

LCOW is great but lingering in limbo. There are many solutions where Windows and Linux containers work together.

**Comment by stephen-turner** [comment]
See #78 for more details.

**Comment by 0x53A** [comment]
For me the important part is just building. Yeah, running both in parallel would be nice, but not strictly required.

Our dockerfiles don't do anything fancy, just copy a folder, set an environment variable and set the entry point.

So I'd also be happy with some other tool. Maybe with WSL2 we can invoke buildkit through WSL or something, idk.

**Comment by thomasoatman** [comment]
Hi.  I am working on moving our build and testing environment to Docker Containers.
We ship Linux and Windows components and thus need support for both.
During testing, we require a Windows client container talking to a database in Linux container.
So the lack of dual support is killing my efforts.
thank you

**Comment by annemartijn0** [comment]
We're bound to windows containers due to the use of 'windows authentication' in some of our services and because of some SLA's. We do use some linux containers like consul and traefik. Not being able to mix Windows and Linux containers prevents our team from starting to use WSL2.

**Comment by conioh** [comment]
@bplasmeijer, @thomasoatman, @annemartijn0: Consider using vanilla Docker instead of Docker Inc.'s broken "Docker Desktop". That's what I'm doing and it solved me this issue along with many others.  I saw a few people have been asking about it so I put the details in https://github.com/conioh/Docker-Windows-Linux so I won't have to repeat myself. I hope this helps you. It's not a perfect replacement, but it goes 99% of the way and that's what I'm using.

**Comment by sirlatrom** [comment]
> 
> 
> @bplasmeijer, @thomasoatman, @annemartijn0: Consider using vanilla Docker instead of Docker Inc.'s broken "Docker Desktop". That's what I'm doing and it solved me this issue along with many others. I saw a few people have been asking about it so I put the details in https://github.com/conioh/Docker-Windows-Linux so I won't have to repeat myself. I hope this helps you. It's not a perfect replacement, but it goes 99% of the way and that's what I'm using.

See https://github.com/docker/roadmap/issues/78#issuecomment-684076615 for a reply to this cross-posted comment.

**Comment by conioh** [comment]
> > @bplasmeijer, @thomasoatman, @annemartijn0: Consider using vanilla Docker instead of Docker Inc.'s broken "Docker Desktop". That's what I'm doing and it solved me this issue along with many others. I saw a few people have been asking about it so I put the details in https://github.com/conioh/Docker-Windows-Linux so I won't have to repeat myself. I hope this helps you. It's not a perfect replacement, but it goes 99% of the way and that's what I'm using.
> 
> See [#78 (comment)](https://github.com/docker/roadmap/issues/78#issuecomment-684076615) for a reply to this cross-posted comment.

See https://github.com/docker/roadmap/issues/78#issuecomment-684078467 for a reply to this bizarre claim.

TLDR: Annoying troll either doesn't understand the difference between the physical network the host is connected to and a private virtual network between the host and a guest VM or he is deliberately misleading and lying.

**Comment by MathiasMagnus** [comment]
I am working on upgrading our CI process to allow testing locally. Both GitHub Actions and GitLab CI don't allow running pipelines locally (of course, they weren't designed to do that). Generally I was advised to move away from the feature set of any CI runner and put as much logic into (shell) scripts as possible. Heeding that advice, I'm working on a CMake module (because all of our build automation is CMake anyway) which is able to run CI tasks locally.

Pipeline definition for GitLab and GitHub remain YAML files which do not much more than define the test matrix, distribute the work to runners and invoke a single script on every pipeline stage. The CMake module is an auxiliary project commited into version control.

- CMake configuration step allows customizing the "local runner", selecting some subset of the test matrix one wishes to test for.
- The build step builds the images (from Dockerfiles from within the repo), ideally in parallel.
- CTest runs the same shell scripts which ordinary CI does (actually, they are CMake scripts too, invoked as `cmake -P Build.cmake` for eg.) in the correct order with dependency tracking inside Docker containers. Tests again ought to run in parallel. (Tests would get labels based on which test matrix elements they inherit, one could select some subset of the test matrix through ordinary CTest CLI args.)

It would be rapid response CI runs without involving remote entities, possibly even run totally offline.

I may be able to orchestrate the entire thing having CMake communicate with two daemons, one locally inside Windows and one running inside WSL2, but setup becomes way more complicated, not to mention having to bind the lifetime of the Linux daemon to a long running Windows process (service?), instead of it being the same daemon which is running on the Windows host anyway.

**Comment by profblackjack** [comment]
My org has .net framework applications I'm looking to containerize, and I would love to be able to mix Linux and windows containers on dev machines for a local development environment more reflective of our deployed environments

**Comment by MarkLFT** [comment]
I also would find it extremely useful to be able to run both Windows and Linux containers on a single host.  We have many scenarios where it is required we run both types of containers in a single installation.

**Comment by derWallace** [comment]
For my org using this feature would help us quite a lot as well.  We have some legacy components which require to mix Linux containers with Windows containers. Manually switching on the Docker Desktop between Linux containers and Windows containers all the time is quite cumbersome for our developers.

**Comment by rompic** [comment]
> For my org using this feature would help us quite a lot as well. We have some legacy components which require to mix Linux containers with Windows containers. Manually switching on the Docker Desktop between Linux containers and Windows containers all the time is quite cumbersome for our developers.

+1

Afaik it is possible via the DockerCLI

**Comment by bplasmeijer** [comment]
@nebuk89 any update?

**Comment by Seramis** [comment]
Any updates?

**Comment by nebuk89** [comment]
Ah sorry for being quiet, no updates yet I am afraid 😞. Please keep everyone following/providing the 👍 so we can see the interest!

**Comment by silverl** [comment]
Another +1 here. We have a mix of Windows .NET Framework services and IIS web sites, and Linux services, and would LOVE to be able to run both on a single dev workstation.

**Comment by bplasmeijer** [comment]
> Ah sorry for being quiet, no updates yet I am afraid 😞. Please keep everyone following/providing the 👍 so we can see the interest!

We need some movement on #79 here, see also https://github.com/microsoft/Windows-Containers/issues/34

**Comment by TBBle** [comment]
The way I read https://github.com/moby/moby/issues/36682#issuecomment-733763007, LCOW itself is not deprecated, merely the current implementation in Docker. containerd has LCOW support already, and there are MS developers still fixing bugs (at least) in the containerd LCOW support since that comment.

It ought to work in Docker without much effort once Docker uses containerd on Windows, because containerd applies LCOW when a container started for Linux in the Windows daemon. So this is a distinct but non-interfering potential solution for this ticket, contrasting with WSL2.

**Comment by conioh** [comment]
Microsoft has something called "LCOW v2". See for example https://github.com/microsoft/opengcs and that's what's being worked on in containerd. This not what you get if you install the 4.14 kernel from https://github.com/linuxkit/lcow (which is archived and no work being done on).

I don't do "once". I do "now". Now WSL2 works (and is supported insofar as you can call any Microsoft product or service "supported").

**Comment by TBBle** [comment]
Oh, thank you, I'd been looking for a reference for "LCOW v2". So in my earlier comment, "the current implementation in Docker" is LCOW v1, and "containerd LCOW support" is LCOW v2, and the rest of the comment still applies.

Note that the v1/v2 split (e.g., visibility of UVM to the HCS API) [also applies to WCOW](https://github.com/microsoft/hcsshim/blob/20ce9b887fbf8d1ee4077b2490b29facf2be0a28/internal/hcsoci/create.go#L334-L352), and [Docker is also relying on containerd integration to support the v2 API for WCOW containers](https://github.com/moby/moby/issues/41455).

And speaking of support, MS has recently been [somewhat explicit that they do _not_ support WSL2 for production workloads](https://github.com/MicrosoftDocs/WSL/issues/678#issuecomment-792294928), only for development, going so far as to [disable it in Windows Server 2022 LTSC prerelease builds](https://techcommunity.microsoft.com/t5/windows-server-insiders/wsl-2-not-working-after-installing-20262/m-p/1928823). So a WSL2-based solution isn't currently viable on Windows Server 2022 LTSC, although I'm not sure Docker desktop is supported on Windows Server anyway.

Since (AFAIK) LCOW v2 doesn't require further OS support than already exists in Windows Containers, I don't know if it will be similarly blocked on Windows Server 2022 LTSC; I hope not, as amongst other things, that'll mess with the CI pipelines for containerd.

**Comment by conioh** [comment]
> Oh, thank you, I'd been looking for a reference for "LCOW v2". So in my earlier comment, "the current implementation in Docker" is LCOW v1, and "containerd LCOW support" is LCOW v2, and the rest of the comment still applies.

It still applies in the sense that it's still a nice prospect for the future but doesn't help people today, on May 2021, to run Linux and Windows containers without wasting years of their lives on "Switch to Windows containers...", "Switch to Linux containers...", "Switch to Windows containers...", "Switch to Linux containers...", "Switch to Windows containers...", "Switch to Linux containers...", "Switch to Windows containers...", "Switch to Linux containers...", "Switch to Windows containers...", "Switch to Linux containers...", "Switch to Windows containers...", "Switch to Linux containers...", "Switch to Windows containers...", "Switch to Linux containers...", ...

Using WSL2 does.

> And speaking of support, MS has recently been [somewhat explicit that they do _not_ support WSL2 for production workloads](https://github.com/MicrosoftDocs/WSL/issues/678#issuecomment-792294928), only for development, going so far as to [disable it in Windows Server 2022 LTSC prerelease builds](https://techcommunity.microsoft.com/t5/windows-server-insiders/wsl-2-not-working-after-installing-20262/m-p/1928823).

So what? Both WSL2 and LCOW (v1, the only thing that exists) are not supported for production. Both. There's no difference in this regard so it's not a disadvantage of using WSL2. But WSL2 works better for development, which is what people here seem to want (building images, not deploying them in production).

> So a WSL2-based solution isn't currently viable on Windows Server 2022 LTSC, although I'm not sure Docker desktop is supported on Windows Server anyway.

It's obviously not. Docker Desktop is only supported and intended to run on Workstation SKUs.

**Comment by TBBle** [comment]
All I said was "LCOW itself is not deprecated", and "it's a distinct but non-interfering potential solution for this ticket, contrasting with WSL2". I guess you're seeing "don't use WSL2 though `docker --context`" somewhere in what I wrote, and responding to that?

**Comment by conioh** [comment]
> All I said was "LCOW itself is not deprecated", and "it's a distinct but non-interfering potential solution for this ticket, contrasting with WSL2". I guess you're seeing "don't use WSL2 though `docker --context`" somewhere in what I wrote, and responding to that?

LCOW - the thing that exists and can be used to run Linux containers - is deprecated.

**Comment by TBBle** [comment]
LCOW (v1), "the experimental thing that exists now in Docker" is deprecated in Docker, yes.

That's the _start_ of this ticket, a request for something in Docker (Desktop I assume) to replicate that nice workflow. LCOW, the concept, and hence LCOW v2 as a possible solution to the use-case of this ticket, is not deprecated, and if it gets delivered, would indeed provide the same workflow.

**Comment by stephen-turner** [comment]
@SetTrend Yes, of course the documentation should reflect the current state. If you can point to a page which is wrong, we'll fix it (or you can open a pull request if you prefer, all our docs are open source).

**Comment by SetTrend** [comment]
I created a number of pull requests already. But no-one cares ... 😢

**Comment by SetTrend** [comment]
On the other hand:

Belonging to the species of professional programmers who try to gain revenue with their own projects - utilizing Docker, perhaps even having a paid Docker subscription plan, only as a tool - I would not commend expecting those users to invest too much of their time in investigating on a tool's roadmap or to update a tool's documentation.

The most important fact for this breed of programmers is what is available here, right now. Not in the future, not in the past. Everything else will sooner or later lead to not recommending the tool.

**Comment by stephen-turner** [comment]
@conioh Your tone is not appropriate and I have deleted your comment.

**Comment by netpoetica** [comment]
> At this moment, I am using the (LCOW) **experimental feature** to run Linux and Windows containers simultaneously.

Is this currently the recommend approach, and is this documented anywhere?

I was hoping that by specifying platform in my docker compose, I would be able to just run both Linux and Windows containers simultaneously. But the Docker Desktop client "Switch to..." option right now I think blocks running one or the other

**Comment by stephen-turner** [comment]
> > At this moment, I am using the (LCOW) **experimental feature** to run Linux and Windows containers simultaneously.
> 
> Is this currently the recommend approach, and is this documented anywhere?
> 
> I was hoping that by specifying platform in my docker compose, I would be able to just run both Linux and Windows containers simultaneously. But the Docker Desktop client "Switch to..." option right now I think blocks running one or the other

Unfortunately LCOW has now been deprecated. It never reached the level of quality where we were happy to mark it as non-experimental, and in the end we dropped it.

There isn't a good solution for running Linux and Windows containers simultaneously right now, which is why this ticket is on the roadmap.

**Comment by silverl** [comment]
Hi, @stephen-turner. I looked on the [roadmap](https://github.com/docker/roadmap) but I don't see this ticket on the board.

If a GitHub Issue remains open (as opposed to WONTFIX or closed), does that imply it's "on the roadmap" but not actively under research or investigation? I.e. The idea hasn't been dismissed, but there are no immediate plans to tackle it.

**Comment by stephen-turner** [comment]
Correct, we're aware of the demand but we're not currently working on it.

**Comment by Scherlac** [comment]
@netpoetica  In my experience in case we start the Windows container and than switch to Linux we can run both at the same time. I mean I have a Cosmos DB running parallel with two Linux hosted DevOps agent ( and a StackEditor image :) )  (Windows 1909, WSL2, Docker Desktop). 
For me it is just inconvenient to switch. The main issue for me is VPN, that prevents network traffic on Windows containers. Linux network traffic is fine, [vpnkit](https://github.com/moby/vpnkit) does a fine job here.

**Comment by aL3891** [comment]
i came across this while looking for ways to run build and run a mixed container solution as we need to use windows containers for some legacy services, but also use Linux to run things like redis. 
I'm a bit surprised this issue hasn't been able to receive more priority since it would unblock so many people from using containers at all.

I did however find a workaround for atleast being able to build windows and Linux containers on the same machine in a non interactive fashion. There is a seemingly undocumented way to switch between Linux and windows containers from the command line detailed [here] (https://stackoverflow.com/questions/57081352/docker-command-line-to-switch-to-running-linux-containers-on-windows-core-machin) and [here](https://forums.docker.com/t/cli-to-switch-between-linux-and-windows-images/30297/10)

Basically you do this before building your windows container
`& $Env:ProgramFiles\Docker\Docker\DockerCli.exe -SwitchWindowsEngine` (or `"%ProgramFiles%\Docker\Docker\DockerCli.exe" -SwitchWindowsEngine` in a cmd file)

and then  before building the linux containers
 `& $Env:ProgramFiles\Docker\Docker\DockerCli.exe -SwitchLinuxEngine`

There are downsides to this approach, switching engine is a global operation so build operations on a build agent must be coordinated so different builds doesn't disrupt each other. I also don't think this solves the issue with actually running containers and having them talk to each other, let alone run with docker compose or kubernetes

Atleast with this approach you don't have to go right clicking into some menu to switch. i don't quite understand why both the linux and windows context can be available at the same time, even if its not displayed in the gui, it would still allow scripts to use the different context

**Comment by travis-icorein** [comment]
> Basically you do this before building your windows container
> `& $Env:ProgramFiles\Docker\Docker\DockerCli.exe -SwitchWindowsEngine` (or `"%ProgramFiles%\Docker\Docker\DockerCli.exe" -SwitchWindowsEngine` in a cmd file)
> 
> and then before building the linux containers
> `& $Env:ProgramFiles\Docker\Docker\DockerCli.exe -SwitchLinuxEngine`
> 
> There are downsides to this approach, Docker desktop must be installed and i'm not sure what the situation is there when running on windows server. Also, switching engine is a global operation so build operations on a build agent must be coordinated so different builds doesn't disrupt each other. I also don't think this solves the issue with actually running containers and having them talk to each other, let alone run with docker compose or kubernetes

Correct, upon executing `& $Env:ProgramFiles\Docker\Docker\DockerCli.exe -SwitchLinuxEngine` anything that existed in the SwitchWindowsEngine environment is no longer visible to the docker system.  Running `docker container ls -a` won't show you containers running in the other environment.  Neither can they communicate (eg: you can't create a network that will be seen by both, and not just merely be two different networks with the same name).

> i don't quite understand why both the linux and windows context can be available at the same time, even if its not displayed in the gui, it would still allow scripts to use the different context

Agreed, this has been an incredible headache to chase down... all to find out that (despite [older claims ](https://devblogs.microsoft.com/premier-developer/running-docker-windows-and-linux-containers-simultaneously/) to the contrary) it doesn't work :-(

...all to try and get a consistent dev environment that works with the [CosmosDb Emulator](https://docs.microsoft.com/en-us/azure/cosmos-db/linux-emulator?tabs=ssl-netstd21#run-on-linux).

**Comment by bplasmeijer** [comment]
Any update @stephen-turner?
Can we expect next steps, or do o need to take  an other road?

**Comment by stephen-turner** [comment]
@bplasmeijer My comment of 8 Jun is still the current situation.

**Comment by bplasmeijer** [comment]
> @bplasmeijer My comment of 8 Jun is still the current situation.

How can we get some movement on this request, containerd does support Linux, and Windows containers at the moment. See also the following thread, and answers, https://twitter.com/Aspenwilder/status/1428432777876107266?s=19

**Comment by slonopotamus** [comment]
> How can we get some movement on this request

You can go and add code that would use containerd functionality to http://github.com/moby/moby

**Comment by TBBle** [comment]
The current working assumption I have is that the only way forward for LCOW in Docker is leveraging the LCOW support in containerd, which I assume was the relevance of that Twitter thread which briefly touched upon the Docker/containerd integration work on Windows.

**Comment by maikschulze** [comment]
Hi,

I've made a quick experiment with Docker contexts and am a surprised that this works. As I'm not very familiar with Docker, I would appreciate if someone could shed light on how this works and what the implications, i.e. limitations or known issues, are, as I have only quickly tested this. I would hope this helps others as well.

I'm running a Windows 10 Professional 20H2 (build 19042) host system with Docker Desktop 3.6.0 installed. Everything was set up to run Linux containers through WSL2 in Linux mode and run Windows containers in isolation mode by setting the Windows mode JSON config `"exec-opts": [ "isolation=process" ]`.

After executing side-by-side containers successfully in Windows mode through `docker --platform=linux run`, I've encountered the issue https://github.com/docker/for-win/issues/10813 which is a sales stopper for me.

I've then stumbled upon this article: https://poweruser.blog/docker-on-windows-10-without-hyper-v-a529897ed1cc which made me aware of the existence of contexts.

On my Windows host, I've created two contexts pointing to endpoints of the managed ones when using `$env:ProgramFiles\Docker\docker\DockerCli.exe -SwitchLinuxEngine` or `$env:ProgramFiles\Docker\docker\DockerCli.exe -SwitchWindowsEngine`:

`docker context create windows --description "Windows containers" --default-stack-orchestrator=swarm --docker "host=npipe:////./pipe/dockerDesktopWindowsEngine"`

`docker context create linux --description "Linux containers" --default-stack-orchestrator=swarm --docker "host=npipe:////./pipe/dockerDesktopLinuxEngine"`

These contexts persist over switching the engine. I've then started containers in two Powershell instances and keep them running simultaneously. At first glimpse, I can see no issues:

Instance 1
```
PS C:\Windows\System32\WindowsPowerShell\v1.0> docker --context linux run --rm -it ubuntu bash
root@277d141d07ca:/# cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.3 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.3 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
root@277d141d07ca:/# exit
exit
PS C:\Windows\System32\WindowsPowerShell\v1.0> $LASTEXITCODE
0
PS C:\Windows\System32\WindowsPowerShell\v1.0>
```

	
Instance 2
```
PS C:\Windows\System32\WindowsPowerShell\v1.0> docker --context windows run --rm -it mcr.microsoft.com/windows/servercore:20H2-amd64 powershell
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

Try the new cross-platform PowerShell https://aka.ms/pscore6

PS C:\> [System.Environment]::OSVersion.Version

Major  Minor  Build  Revision
-----  -----  -----  --------
10     0      19042  0


PS C:\> exit
PS C:\Windows\System32\WindowsPowerShell\v1.0> $LASTEXITCODE
0
PS C:\Windows\System32\WindowsPowerShell\v1.0>
```


Context-specific commands such as
`docker --context linux container ls`

appear to work. It seems, this is a viable solution to run WSL2 and Windows isolation containers side-by-side. Am I missing something?

Best regards
Maik

**Comment by slonopotamus** [comment]
> Am I missing something?

Can these containers talk over network to each other?

**Comment by maikschulze** [comment]
> Can these containers talk over network to each other?

Yes, that seems to work. I have started Python web servers in the Linux and Windows containers and exposed their ports. The containers can connect to the other container's server and are reachable through localhost from the host. I cannot find any difference in the behavior in comparison to using the actual "Switch to Linux/Windows containers" in Docker Desktop. What also worked what elevating rights through `--cap-add=NET_ADMIN` for the Linux container.

I have also tested volumes and mounts successfully.

The behavior was identical in both Linux and Windows mode. I'm unsure what the switch actually serves apart from setting the default docker context, e.g. `docker context use <context>`, and some GUI setting in Docker Desktop.

For completeness, you can avoid the explicit creation of the contexts and use the environment variable `DOCKER_HOST` to point to npipe and get the same result, see https://docs.docker.com/engine/reference/commandline/cli/#environment-variables

**Comment by ohault** [comment]
Will the switching setting in Docker Desktop for Windows be removed soon in a future release (5.0)?

**Comment by tomatac** [comment]
Is running a mix of Windows and Linux containers from a docker-compose on a Widows host stable/supported?

**Comment by christophermclellan** [comment]
Hi folks, just to reiterate on @stephen-turner's comments from 20-August 2021 and 08-June 2021, this is not something that we are working on atm or currently have planned in on our roadmap.

**Comment by abehartbg** [comment]
For anyone who runs across this @maikschulze solution will probably work for most use cases. Only thing I've noticed is that both Linux and Windows need to be 'switched to' at least once to have the corresponding file handles available. I do this with this command `& $env:ProgramFiles\Docker\docker\DockerCli.exe -SwitchLinuxEngine`

**Comment by maikschulze** [comment]
Great to  hear this is working for you @abehartbg . I've made a small startup script sequence where I switch between the two engines and wait a few seconds in between.
Ever since, I've been using this approach successfully in a CI production environment. In Gitlab you can configure this setup easily:
```
[[runners]]
  name = "<LINUX-NAME>"
  executor = "docker"
  ..
  [runners.docker]
    host = "npipe:////./pipe/docker_engine_linux"
    ..

[[runners]]
  name = "<WINDOWS-NAME>"
  executor = "docker-windows"
  ..
  [runners.docker]
    host = "npipe:////./pipe/docker_engine_windows"
    ..
```

**Comment by TBBle** [comment]
This is because it's the Docker Desktop intervening proxy (the thing at `npipe:////./pipe/docker_engine`) that rewrites Windows paths to work with the Linux engine before passing the request on to `npipe:////./pipe/docker_engine_linux` when in Linux Containers mode, or pretty-much just passes it straight through to `npipe:////./pipe/docker_engine_windows` when in Windows Containers mode.

You probably _could_ make

```
$env:DOCKER_HOST="npipe:////./pipe/docker_engine_linux"
docker run --rm -it -v c/:/windows_mount alpine ls /windows_mount
```

work, but using a WSL2-visible path instead, e.g., something like

```
$env:DOCKER_HOST="npipe:////./pipe/docker_engine_linux"
docker run --rm -it -v /mnt/c/:/windows_mount alpine ls /windows_mount
```

assuming the WSL2 backend. If that works, it'd also work if Docker Desktop was in Windows Containers mode, since using that `DOCKER_HOST` value, you're bypassing Docker Desktop entirely.

Note that I haven't checked this, it's possible that the WSL2 backend setup in Docker Desktop relies on Docker Desktop to do other things than rewrite the paths in order to mount volumes from Windows.

**Comment by maikschulze** [comment]
Thank you very much for this information!

I just tried in Windows mode and Linux mode to mount like this
```
$env:DOCKER_HOST="npipe:////./pipe/docker_engine_linux"
docker run --rm -it -v /mnt/c/:/windows_mount alpine ls /windows_mount
```
The container starts but no files or directories are listed. I would imagine that in addition to the path rewrite, some kind permission management is also done.

**Comment by TBBle** [comment]
From memory (I poked deep into it using nsenter at one point), the dockerd process in WSL2 is running inside container, and so the relevant mount point where the Windows partitions are mounted might not be `/mnt/c/`, and it may be that as you suppose, the Windows partitions are _not_ mounted inside that container until Docker Desktop sees that a path is needed by a mount request it's rewriting.

That'd be a good security practice, in fact.

**Comment by andrewtek** [comment]
I am supporting some legacy ASP.NET applications that use SOLR. I would like to create a single docker-compose that uses windows containers for the ASP.NET components and Linux for the SOLR components.

I have done some experiments running Linux containers from WSL2 while running Windows containers from Visual Studio while Docker Desktop is in Windows mode. I can access both containers (yeah), but I want a single docker-compose that spins up my entire environment with a single button-press.

Any chance this will happen?

**Comment by mloskot** [comment]
@andrewtek 
>  I want a single docker-compose that spins up my entire environment

Have you experimented with this by any chance?
https://devblogs.microsoft.com/premier-developer/mixing-windows-and-linux-containers-with-docker-compose/

**Comment by thaJeztah** [comment]
Oh! Sorry I now see I `@`-mentioned the wrong person (sorry for that) 🙈

**Comment by andrewtek** [comment]
Thanks for the update @mloskot  and @thaJeztah . I did try out that article, but it did not work for me. As @thaJeztah said, the feature was deprecated. 

For my situation, I found that I could not run SOLR using LCOW, but it runs great using WSL2. I love that SOLR has done so much to help the community with pre-made images and documentation to get up and running,
https://solr.apache.org/guide/solr/latest/deployment-guide/solr-in-docker.html

Unfortunately, if you need to work on an ASP.NET Framework application, you cannot benefit from those SOLR images. Is there a technical reason why the deprecated feature cannot be re-implemented with WSL2?

**Comment by TBBle** [comment]
I don't have the API specs in front of me, but if the Docker API had something that could identify the platform of a particular request, e.g., like the `--platform` flag for old LCOW, then a proxy (such as Docker Desktop, *cough*) could be used to split such requests between the native Windows dockerd and the WSL2-hosted dockerd. Right now Docker Desktop switches mode to support this same split one at a time, but both daemons are running.

I also confess I'm not really famliar with Docker Compose, can it be told that different parts of the system run on different dockerd instances, automating some of the ideas earlier in this thread? I'm not sure if it'd be feasible for, e.g., a backend private network shared between a Windows container and a Linux (WSL2-hosted) container, but I'm not sure it wouldn't be doable. (It also might be more complexity than just waiting for "New" LCOW to get into people's hands; my understanding is that it's _workable_ on the containerd side now, but for Docker users it also depends on containerd-on-Windows and also further front-end work to enable it; I also haven't tried it myself, so might be overassuming its current state)

**Comment by davhdavh** [comment]
Why not just start both the windows container mode and wsl2 container mode at the same time?
Then automatically configure the two in a swarm with some appropriate labels and 99% of the problems goes away.
Then if i docker run from wsl, it will talk to the linux engine, and if I docker run from powershell it will talk to windows engine, and docker-compose can use the labels (or platform:, or just image) to start on the proper engine.

**Comment by TBBle** [comment]
You can do roughly that now, using Docker contexts per https://github.com/docker/roadmap/issues/79#issuecomment-1002424911 (and see https://github.com/docker/roadmap/issues/79#issuecomment-1289673584 and the following few comments for caveats and notes). I'm not sure if Docker compose can work across contexts though, or just talks to the current context selected with `docker context use`.

**Comment by davhdavh** [comment]
Yes, that was kinda my point. It just needs to be the default or a checkbox to enable. That will send a clear signal to other tools to properly support it also

**Comment by wizpresso-steve-cy-fan** [comment]
We are also interested to see if LCOW can be upgraded to support WSL2 as a backend. 

The majority of our container applications right now is Windows based, but some of GPU workloads are based on Linux Container. 

We can also confirm that [#79](https://github.com/docker/roadmap/issues/79#issuecomment-1335118698) worked to some extent but we don't know how to keep both Windows and Linux Docker daemon running simultaneously without using `& $env:ProgramFiles\Docker\docker\DockerCli.exe -SwitchWindowsEngine` and `& $env:ProgramFiles\Docker\docker\DockerCli.exe -SwitchLinuxEngine`

PS: LCOW support is deprecated but not entirely remove AFAIK, and you can still turn that on with `"experiment": true`. Obviously however, there doesn't seems to have GPU support unlike WSL2 docker.

**Comment by TBBle** [comment]
The old LCOW setup is removed from Docker in 23.0, and was already broken in various ways for years before that.

The new LCOW approach (Sometimes "LCOW v2", but I'm not sure if that's ever been used in a formal sense) is actually the same underlying tech as WSL2 (micro-VMs), but getting it into Docker depends on getting Windows Docker to be based on containerd, which has been a multi-year project already: https://github.com/moby/moby/issues/41455 which I've apparently already referenced here in 2021, and others linked from there. (Also, I've lost track of what outstanding LCOW-specific work is still TBD in containerd itself. I believe it's been working in branches, but I'm not sure if they have all been merged into containerd yet.)

Once that's done, we will be pretty close (in a tech-stack sense, not in the "rainy afternoon's hacking" sense) to getting what this ticket wants: because containerd doesn't have "Windows" and "LCOW" modes, each container (or sandbox, really) can use a different runtime. Inter-container behaviour between LCOW and Windows containers will probably still have limitations, but being able to build Linux and Windows images in a single Docker daemon without changing modes would only require work in Docker itself, e.g. the UX could be reinstating the old (LCOW) `--platform` parameter for build and execution commands, but the implementation would be quite different underneath.

This might also open the door for Compose using a mix of platforms (which I don't think worked with LCOW v1), but I don't know enough about Compose to say anything further on that topic.

**Comment by linghengqian** [comment]
That being said, I don't see any issues tracking WSL2 at https://github.com/containerd/containerd/issues . Does this mean that the current issue is actually still blocked by https://github.com/moby/moby/issues/41455 on the moby side?

**Comment by slonopotamus** [comment]
Are there any docs (even basic) on how to use LCOW in containerd? Like, which runtime shim to use, where to get kernel, where to put it, etc.
